File: kotlin/src/main/kotlin/765.CouplesHoldingHands.kt
import org.junit.Assert.assertEquals
import org.junit.Test
//https://leetcode.com/problems/couples-holding-hands/


class CouplesHoldingHands{
  fun minSwapsCouples(row: IntArray): Int {
    val index = hashMapOf<Int, Int>()
    for((i, v) in row.withIndex())
      index[v] = i

    var count = 0
    for(i in 0 until row.size step 2){
      val a = row[i]
      val b = a xor 1
      if(index[b] != i+1){
        count ++
        index[row[i+1]] = index[b]!!
        row[index[b]!!] = row[i+1]
      }
    }
    return count
  }

  @Test
  fun test1(){
    assertEquals(1, minSwapsCouples(intArrayOf(0, 2, 1, 3)))
  }

  @Test
  fun test2(){
    assertEquals(0, minSwapsCouples(intArrayOf(3, 2, 0, 1)))
  }
}

File: python/765_CouplesHoldingHands.py
from unittest import TestCase
# https://leetcode.com/problems/couples-holding-hands/


class CouplesHoldingHands(TestCase):
    def minSwapsCouples(self, row: 'List[int]') -> int:
        index = dict()
        for i,v in enumerate(row):
            index[v] = i

        count = 0
        for i in range(0, len(row), 2):
            a, b = row[i], row[i+1]
            aa = row[i] ^ 1
            if aa != b:
                count += 1
                row[index[aa]] = b
                index[b] = index[aa]
        return count

    def test1(self):
        self.assertEqual(1, self.minSwapsCouples([0, 2, 1, 3]))

    def test2(self):
        self.assertEqual(0, self.minSwapsCouples([3, 2, 0, 1]))
